### ISSUE TITLE
swarm not functioning when one of two swarm managers is down

### ISSUE THREAD

**Issue by ligc** [docker/docker#24407]
**Output of `docker version`:**

```
root@c910f04x19k05:~# docker version
Client:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   906eacd-unsupported
 Built:        Fri Jun 17 21:21:56 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   906eacd-unsupported
 Built:        Fri Jun 17 21:21:56 2016
 OS/Arch:      linux/amd64
root@c910f04x19k05:~# 
```

**Output of `docker info`:**

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 1.12.0-rc2
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 22
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay bridge null host
Swarm: active
 NodeID: eli8zcc9qkvpyymwnkbmhcphp
 IsManager: Yes
 Managers: 2
 Nodes: 4
 CACertHash: sha256:424c3732f099885c59d66c9a498e4d8a99ace8e9186d561117a58f807b46b31d
Runtimes: default
Default Runtime: default
Security Options: apparmor seccomp
Kernel Version: 4.4.0-22-generic
Operating System: Ubuntu 16.04 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.859 GiB
Name: c910f04x19k05
ID: TJGZ:GRAN:BCTW:5WN7:AA44:EDV4:RS4M:2RV6:RXU7:XOHR:76XK:KETQ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Steps to reproduce the issue:**
1. Setup a docker swarm with 2 managers
2. Deploy services in the swarm
3. Shutdown one of the manager
4. You could not manipulate the service, node or swarm 

**Describe the results you received:**
```
root@c910f04x19k05:~# docker node ls            
ID                           NAME           MEMBERSHIP  STATUS  AVAILABILITY  MANAGER STATUS
7ciwf38osn5u05cyktcsqi3ez    c910f04x19k06  Accepted    Ready   Active        
7x1cdxcttssryr5r7ock1zx7c    c910f04x19k07  Accepted    Ready   Active        
eke7jq96u04yqxdt8z4vyrm9q    c910f04x19k04  Accepted    Ready   Active        Unreachable
eli8zcc9qkvpyymwnkbmhcphp *  c910f04x19k05  Accepted    Ready   Active        Leader
root@c910f04x19k05:~# docker node rm c910f04x19k04 =========> Could not delete the failed manager
Error response from daemon: rpc error: code = 9 desc = node eke7jq96u04yqxdt8z4vyrm9q is a cluster manager and is part of the quorum. It must be demoted to worker before removal
root@c910f04x19k05:~# docker node demote c910f04x19k04
Error response from daemon: rpc error: code = 4 desc = context deadline exceeded
root@c910f04x19k05:~# docker node rm c910f04x19k04    
Error response from daemon: rpc error: code = 9 desc = node eke7jq96u04yqxdt8z4vyrm9q is a cluster manager and is part of the quorum. It must be demoted to worker before removal
root@c910f04x19k05:~# 

root@c910f04x19k05:~# docker service tasks sshd
ID                         NAME    SERVICE  IMAGE                             LAST STATE          DESIRED STATE  NODE
5ggkrqm5ya0lajzsqup4p8hsj  sshd.1  sshd     liguangcheng/ubuntu-16.04-x86_64  Running 18 minutes  Running        c910f04x19k04
dziaej6obfxwchgld0ejfu5n4  sshd.2  sshd     liguangcheng/ubuntu-16.04-x86_64  Running 22 minutes  Running        c910f04x19k07
dokmapfyo9erx4sncfkkfxe7l  sshd.3  sshd     liguangcheng/ubuntu-16.04-x86_64  Running 22 minutes  Running        c910f04x19k05
0a7c7brujzdedljv4dtdf4jt6  sshd.4  sshd     liguangcheng/ubuntu-16.04-x86_64  Running 22 minutes  Running        c910f04x19k06
4dkb78sool9lncm8dd8ivyazw  sshd.5  sshd     liguangcheng/ubuntu-16.04-x86_64  Running 22 minutes  Running        c910f04x19k04
root@c910f04x19k05:~# docker service rm sshd =========> Could not remove the service
Error response from daemon: rpc error: code = 4 desc = context deadline exceeded
root@c910f04x19k05:~# docker service create --replicas 5 --publish 22 --name sshd-1 liguangcheng/ubuntu-16.04-x86_64 /usr/sbin/sshd -D ===========> Could not launch new services
Error response from daemon: rpc error: code = 4 desc = context deadline exceeded
root@c910f04x19k05:~# 



```

**Describe the results you expected:**
I should be able to delete the failed manager from the swarm, and be able to manipulate the services in the swarm.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Comment by thaJeztah** [comment]
Could this be a duplicate of https://github.com/docker/docker/issues/24329?

**Comment by ligc** [comment]
Could be, #24329 reads "The only solution is to remove all members from the swarm then redo everything", but the problem for me is that I could not remove the failed manager, so I could not destroy the swarm and redo everything. If there is any force method to destroy the swarm and start from the beginning, it will be fine for me.

**Comment by rmdashrfslash** [comment]
I had this same issue.  destroyed my swarm by doing the following:
* shut down docker
* rm -fr /var/lib/docker/*
* delete and recreate thin storage (If you're using it)
* start docker

**Comment by cpuguy83** [comment]
Yes, you should not setup with only 2 managers.
If either manager goes down you will lose quorum and not be able to recover.

If you have a cluster with 1 manager, and the manager goes down the cluster goes down, but you can bring it back up by fixing the node.

If you have a cluster with 2 managers, **you cannot tolerate any failure**.

If you have a cluster with 3 managers, you can tolerate a single manager failure and it will continue working just fine.

If you need some level of fault tolerance, then you need at least 3 managers.
Basically you can tolerate up to (N/2)+1 failures from 3 nodes and up.
An even number is going to cause problems.

It would be better to have N-1 managers (where N is the current number of managers) instead of an even number.

**Comment by tonistiigi** [comment]
With `rc4` you should be able to demote 2 manager cluster to single manager cluster. Everything that @cpuguy83 said still applies on failure cases.

**Comment by thaJeztah** [comment]
Ah, missed that it was reported with only two managers. I'm closing this issue, because demoting from two to one manager is already tracked in https://github.com/docker/docker/issues/23903, and will be in RC4 as @tonistiigi mentioned above.

Feel free to comment though!

**Comment by adrecord** [comment]
> Basically you can tolerate up to (N/2)+1 failures from 3 nodes and up.

I believe this should say you can tolerate up to (N/2)-.5 failures for an odd number of nodes.  (Obviously don't use an even number.)

I realize this issue is closed... just commenting in case somebody else runs across this.
